Feature/useful contacts card

franciscopana opened the pull request:
Closes #396
We've created the useful contacts page, accessible by the navigation drawer, and filled it with the contacts of academic services and infodesk.

# Review checklist
- Page is available on navigation drawer
- Each topic (Infodesk and Academic Services) has its own card on the page
- Email adresses and phone numbers are still not clickable (that's not the purpose of this issue)
- We've kept text align to left on the cards because of the results of the poll in the project channel

<img width="1440" alt="Screenshot 2022-02-17 at 14 23 41" src="https://user-images.githubusercontent.com/82325571/154501674-22ffc45b-dc47-4989-91b0-fc7a7e085b39.png">

andrefmrocha commented:
You have committed the .idea files, can you remove them?

srissirs reviewed (commented):
Hey! Overall, it looks good to me! Although I would suggest maybe adding a title to some of those informations such as "Horários: " for all the "Atendimento" info, and maybe something like "Áreas: " or "Departamentos: " before all the emails in the Academic Services section. 

It might improve a bit in terms of organization, though I'm not sure if the page might also become too cluttered...

franciscopana commented:
Well, I understand your suggestions and agree with them. In fact, that would clarify better the meaning of the information in the card. I've made some changes in order to implement those 'title labels' and this is the result:
![Screenshot_1645659681](https://user-images.githubusercontent.com/82325571/155429732-faa79b86-300f-4027-895a-0d59c174c513.png)

I'm a little afraid that now there's too much information in the card... please tell me your opinion :)

srissirs commented:
Hum...Could you maybe try to increase the card's height, so that the space between the information remains the same as in the first version? Just to see if that makes it less cluttered.

franciscopana commented:
hmmm I didn't understand very well... We can, basically, divide this new structure in 3 blocks (Horários, Telefone and Departamentos)
Are you suggesting to increase the space between those 3 blocks? Or do you mean more spaced information inside each of those blocks?

srissirs commented:
I meant the info below "Departamentos". The other blocks seem well spaced to me.

franciscopana commented:
So, here it is! The topics inside each block are a bit more spaced and there's more space between the main blocks, too. 
I like it, what's your opinion? We can still add some more pixels, but it may be too much.

<img src="https://user-images.githubusercontent.com/82325571/155586217-399a4c1c-e8f5-4669-a59d-911d41de693d.png" width="300" />

srissirs commented:
It looks really good! Maybe now we could just make the size of the titles in the infoDesk card the same size as the titles in the Academic Services' card? Or do you think we should leave it as it is?

franciscopana commented:
We've used bigger and smaller titles to demarcate a "little hierarchy" between them... I mean, grouping them according to their topics. But the Infodesk card doesn't have so much information as the academic services one: It's rather more concise and simple. So I don't know if changing titles' font size would be a good idea.
But I'll change it, just for testing purposes, and then we'll decide which style we should adopt.

franciscopana commented:
So, this is how Infodesk card looks with bigger titles:

<img src="https://user-images.githubusercontent.com/82325571/156259494-bd58f85c-1da9-4459-9bad-b6a9829c6f9a.png" width="300" />

Well, it looks better than what I was expecting... But I have another proposal:

<img src="https://user-images.githubusercontent.com/82325571/156260488-22f01212-8396-4cd2-a97a-5458808d495e.png" width="300" />

This is, to me, the best option. What do you think?

srissirs commented:
I really like the last option! I wouldn't change anything now. :)

isabel-amaral commented:
Hey! I really like how the design of the cards is looking after these changes. There's just one small detail I'd like to add regarding the navigation drawer. I think here it would be more organized if the "Contactos Úteis" navigator was placed after the "Autocarros" one and before "Sobre".

![Captura de ecrã 2022-03-07 004831](https://user-images.githubusercontent.com/80794891/157084698-dc62ed69-f56a-41d5-acab-4bfbf3482b0f.png)

isabel-amaral commented on app_feup/lib/view/Widgets/info_desk_card.dart:
```diff
@@ -0,0 +1,100 @@
+      : super.fromEditingInformation(key, editingMode, onDelete);
+
+  @override
+  Widget buildCardContent(BuildContext context) {
+    return Column(children: [
+      Table(
+          defaultVerticalAlignment: TableCellVerticalAlignment.middle,
+          children: [
+            TableRow(children: [
+              Container(
+                margin: const EdgeInsets.only(
+                    top: 20.0, bottom: 0.0, left: 20.0), // add margin left: 20
+                child: Text('Horários',
+                    //textAlign: TextAlign.center,
+                    style: Theme.of(context)
+                        .textTheme
+                        .headline4
+                        .apply(fontSizeDelta: 0)),
+              )
+            ]),
+            TableRow(children: [
+              Container(
+                margin: const EdgeInsets.only(
+                    top: 13.0, bottom: 0.0, left: 20.0), // add margin left: 20
+                child: Text('Atendimento presencial e telefónico',
+                    //textAlign: TextAlign.center,
+                    style: Theme.of(context)
+                        .textTheme
+                        .headline4
+                        .apply(fontSizeDelta: -4)),
+              )
+            ]),
+            TableRow(children: [
+              Container(
+                  margin:
+                      const EdgeInsets.only(top: 8, bottom: 12.0, left: 20.0),
+                  child: Text('9:30h - 13:00h | 14:00h - 17:30h',
+                      //textAlign: TextAlign.center,
+                      style: Theme.of(context).textTheme.headline3))
+            ]),
```
Personally, I think the widgets organization is a bit confusing. Instead of using a Table inside a Column and a set of TableRow's for the lines containing the necessary info, I think it would be more simple and readable to just use a Column of Containers. For exemple, something like this (which can also be done in the academic_services_card). What do you think?

```suggestion
  Widget buildCardContent(BuildContext context) {
    return Column(
      children: <Container> [
        Container(
          margin: const EdgeInsets.only(top: 20.0, bottom: 0.0, left: 20.0),
          child: Align(
            alignment: Alignment.centerLeft,
            child: Text('Horários',
                style: Theme.of(context)
                    .textTheme.headline4
                    .apply(fontSizeDelta: 0)),
          ),
        ),
        Container(
          margin: const EdgeInsets.only(top: 13.0, bottom: 0.0, left: 20.0),
          child: Align(
            alignment: Alignment.centerLeft,
            child: Text('Atendimento presencial e telefónico',
                style: Theme.of(context)
                    .textTheme.headline4
                    .apply(fontSizeDelta: -4)),
          ),
        ),
        Container(
          margin: const EdgeInsets.only(top: 8, bottom: 12.0, left: 20.0),
          child: Align(
            alignment: Alignment.centerLeft,
            child: Text('9:30h - 13:00h | 14:00h - 17:30h',
                style: Theme.of(context)
                    .textTheme.headline3),
          ),
        ),
      ],
```

isabel-amaral reviewed (commented):
I also have a few comments regarding the code itself but keep in mind that I'm still learning as well, so if you disagree with anything let me know! :)
And don't forget to remove the .idea files

franciscopana commented:
Thank you for your suggestions! I agree with almost all of them. In fact, the new placement of 'contactos úteis' in the navigation drawer makes sense. The 'context' argument was unuseful, and changing the card from a table with several rows to a column with containers is a great improvement for the code's organization. I just may need some help with the .idea files... honestly, I don't have an idea of what the problem is, maybe because I still don't understand very well what they are.

srissirs reviewed (commented):
In terms of design I think everything is settled. I really like the final version!
In terms of code, it also looks good! 
I would just suggest removing the commented lines inside the `Container` blocks  [here](app_feup/lib/view/Widgets/academic_services_card.dart) and [here](app_feup/lib/view/Widgets/info_desk_card.dart) before merging, since they don't seem to do anything.
But other than that, I think it's ready to merge!

franciscopana commented:
Well, as you can see I've deleted those commented instructions and updated the title of infodesk card from "Info Desk" to "Infodesk". That's how FEUP calls that service in their emails and website. Do you agree?

However, I've some bad news. Firstly, this is the webpage on which we've based the information for our page:
https://sigarra.up.pt/feup/pt/web_base.gera_pagina?p_pagina=*contactos

The problem is that we've found two more pages with more information... and some of them are ambiguous. I mean, the same service has different email addresses and phone numbers. You can see them here:
https://sigarra.up.pt/feup/pt/web_base.gera_pagina?p_pagina=contactos
https://sigarra.up.pt/feup/pt/web_base.gera_pagina?p_pagina=contactos%20portal%20alumni

What shall we do? I and Gonçalo think we should filter the most important information and even attend at infodesk in order to clarify this problem directly with them... but what's your opinion?

Ah, just one more problem... after my last 'git pull', the title of the Academic Services card appears divided into two lines... That's not supposed to happen and I honestly don't know why it's happening. But that's something we still have to fix

isabel-amaral commented:
> Well, as you can see I've deleted those commented instructions and updated the title of infodesk card from "Info Desk" to "Infodesk". That's how FEUP calls that service in their emails and website. Do you agree?
> 
> the title of the Academic Services card appears divided into two lines... That's not supposed to happen and I honestly don't know why it's happening. But that's something we still have to fix


I agree with the new name for the card and I think overall everything is looking good! I checked to see if I could find the problem with the Academic Services card title and it looks like it was caused by some change in the Generic Card class definition that was recently merged, although I don’t understand why either  :confused:
Tbh, I don’t think that it makes the page look worse… but if you prefer the way it was before, maybe it’s worth trying to find out why this is happening...

isabel-amaral commented:
> I've some bad news. Firstly, this is the webpage on which we've based the information for our page: https://sigarra.up.pt/feup/pt/web_base.gera_pagina?p_pagina=*contactos
> 
> The problem is that we've found two more pages with more information... and some of them are ambiguous. I mean, the same service has different email addresses and phone numbers. You can see them here: https://sigarra.up.pt/feup/pt/web_base.gera_pagina?p_pagina=contactos https://sigarra.up.pt/feup/pt/web_base.gera_pagina?p_pagina=contactos%20portal%20alumni
> 
> What shall we do? I and Gonçalo think we should filter the most important information and even attend at infodesk in order to clarify this problem directly with them... but what's your opinion?

I also did some research and found this page:
https://paginas.fe.up.pt/~ipoint/bem-estar/intelectual/servicos-uteis-na-feup/ 
which features the same contacts as in the last two links you sent. 
If you look carefully at the top of page you used, you can see that it’s inserted in ‘Portal do Candidato’ and, judging by the email address, it looks like this subdivision of Infodesk is more for providing information for people wanting to apply to FEUP. Since these people won’t have access to uni I think it doesn’t make sense to include these contacts here, right? It could be replaced by the ones present in the link I included above:

9h30 – 13h00 / 14h00 – 17h30
+351 225 081 400
infodesk@fe.up.pt

srissirs commented on app_feup/lib/view/Widgets/copy_center_card.dart:
```diff
@@ -0,0 +1,33 @@
+class CopyCenterCard extends GenericCard {
+  CopyCenterCard({Key key}) : super(key: key);
+
+  CopyCenterCard.fromEditingInformation(
+      Key key, bool editingMode, Function onDelete)
+      : super.fromEditingInformation(key, editingMode, onDelete);
```
I'm not sure this function is really necessary. Since this card can't be added to the main page, it can't really be edited by the user. I'm not aware of any plans to make these cards editable in the future, but if that's not the case this function can be removed.
The same goes for all the cards on this page.
(Bear in mind that I'm only just starting to use Flutter, so if this comment doesn't make sense feel free to just ignore it :upside_down_face: )

franciscopana replied:
For now, we aren't planning to make this page editable. So, in fact, that function seems useless. 
I'm not very acquainted with Flutter too, but I think we can remove it. When we started working on this issue we copied another card from the main page and that's probably the reason why those lines of code exist... We've never deleted them. Let's do it now :)

srissirs reviewed (commented):
Also, would it be possible to put the card titles in one single line? It looks a bit strange to have some titles occupying two lines.
Other than that everything looks great! I really like these new changes. The information seems far more pertinent to Feup students  :relaxed:

franciscopana commented:
Great! And yes, we'll work on the titles, they are not supposed to be split into 2 lines if they can appear in only 1... But we will do it in a new issue. It's probably a general problem related to the structure of the cards after the addition of the drag icon... So after closing this issue, we are planning to make the contacts clickable and fix the disposition of the titles.

srissirs reviewed (dismissed):
Looks good to me! :ok_hand:

isabel-amaral commented on app_feup/lib/view/Widgets/info_desk_card.dart:
```diff
@@ -0,0 +1,65 @@
+import 'package:flutter/material.dart';
+import 'generic_card.dart';
+
+Container h1(String text, BuildContext context, {bool initial = false}) {
+  final double marginTop = initial ? 15.0 : 30.0;
+  return Container(
+      margin: EdgeInsets.only(top: marginTop, bottom: 0.0, left: 20.0),
+      child: Align(
+        alignment: Alignment.centerLeft,
+        child: Text(text,
+            style:
+                Theme.of(context).textTheme.headline4.apply(fontSizeDelta: 0)),
+      ));
+}
+
+Container h2(String text, BuildContext context) {
+  return Container(
+      margin: const EdgeInsets.only(top: 13.0, bottom: 0.0, left: 20.0),
+      child: Align(
+        alignment: Alignment.centerLeft,
+        child: Text(text,
+            style:
+                Theme.of(context).textTheme.headline4.apply(fontSizeDelta: -4)),
+      ));
+}
+
+Container infoText(String text, BuildContext context, {bool last = false}) {
+  final double marginBottom = last ? 8.0 : 0.0;
+  return Container(
+      margin: EdgeInsets.only(top: 8, bottom: marginBottom, left: 20.0),
+      child: Align(
+        alignment: Alignment.centerLeft,
+        child: InkWell(
+          child: Text(
+            text,
+            style: Theme.of(context).textTheme.headline3,
+          ),
+        ),
+      ));
+}
```
I think it was a great idea to create these functions and reuse them in every card! However, looking at the code's organization, I think it would make more sense if these functions were defined in the [useful_contacts_page](https://github.com/NIAEFEUP/project-schrodinger/blob/feature/useful-contacts-card/app_feup/lib/view/Pages/useful_contacts_card_page_view.dart) (outside the class definition) instead of in the infodesk card file since it's the page that is the "parent" widget of the other cards.

isabel-amaral commented on app_feup/lib/view/Pages/useful_contacts_card_page_view.dart:
```diff
@@ -0,0 +1,28 @@
+}
+
+/// Manages the 'Useful Contacts' section of the app.
+class UsefulContactsCardViewState extends GeneralPageViewState {
+  @override
+  Widget getBody(BuildContext context) {
+    return ListView(children: childrenList());
+  }
+}
+
+List<Widget> childrenList() {
+  final List<Widget> list = [];
+  list.add(InfoDeskCard());
+  list.add(DonaBiaCard());
+  list.add(CopyCenterCard());
+  list.add(MulimediaCenterCard());
+  return list;
+}
```
The childrenList function is only used as an helper funtion to create the structure of the page. Therefore I think it makes sense to add it to the class as an auxiliary method.
```suggestion
class UsefulContactsCardViewState extends GeneralPageViewState {
  @override
  Widget getBody(BuildContext context) {
    return ListView(children: childrenList());
  }
  
  List<Widget> childrenList() {
    final List<Widget> list = [];
    list.add(InfoDeskCard());
    list.add(DonaBiaCard());
    list.add(CopyCenterCard());
    list.add(MulimediaCenterCard());
    return list;
  }
}
```

isabel-amaral commented:
Other than those two small details I think it's good to go! So after you make the changes don't forget to also update the pubspec and changelog 😄

franciscopana commented:
Great suggestions, thank you! I think now everything is ready! 😄

isabel-amaral commented:
Everything looks great! Pls update the branch and I'll aprove this

srissirs reviewed (approved):
Ready to go! :steam_locomotive:
